File: 241DifferentWaystoAddParentheses/LC241.java
class Solution {
    Map<String,List<Integer>> memo = new HashMap<>();
    public List<Integer> diffWaysToCompute(String expression) {
        List<Integer> res = new ArrayList<>();
        if (memo.containsKey(expression)) {
                return memo.get(expression);
        }
        
        if (!containOperator(expression)){
            res.add(Integer.parseInt(expression));
            return res;
        }
        
        for (int i = 0; i < expression.length(); i++) {
            char ch = expression.charAt(i);
            if (ch == '+' || ch == '-' || ch == '*') {
                List<Integer> left = diffWaysToCompute(expression.substring(0,i));
                List<Integer> right = diffWaysToCompute(expression.substring(i+1));
                for (int l: left){
                    for (int r: right) {
                        switch(ch) {
                            case '+':
                                res.add(l+r);
                                break;
                            case '-':
                                res.add(l-r);
                                break;
                            default: //*
                                res.add(l*r);
                            
                        }
                    }
                }
                memo.put(expression, res);
                
            }
            
        }
        //memo.put(expression, res);
        return res;
     
    }
    
    
    public boolean containOperator(String expression) {
        if (expression.length() == 0){
            return false;
        } 
        for(char i: expression.toCharArray()) {
            if(i == '+'|| i=='-' || i == '*') {
                return true;
            }
        
        }
        return false;   
    }
    
    
}

File: 241DifferentWaystoAddParentheses/LC241.py
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        if expression.isdigit():
            return [eval(expression)]
        res = []
        for i,s in enumerate(expression):
            if s in "+-*":
                left = self.diffWaysToCompute(expression[:i])
                right = self.diffWaysToCompute(expression[i+1:])
                for l in left:
                    for r in right:
                        if s == '+':
                            res.append(l+r)
                        elif s == '-':
                            res.append(l-r)
                        else:# s= '*'
                            res.append(l*r)
                            
        return res

# with memozation
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
       
        
        def dfs(expression,memo):
            res = []
            if expression in memo:
                return memo[expression]
            if expression.isdigit():
                return [eval(expression)]

            for i,s in enumerate(expression):
                if s in "+-*":
                    left = dfs(expression[:i],memo)
                    right = dfs(expression[i+1:],memo)
                    res.extend(eval(str(l)+s+str(r))for l in left for r in right)

                    memo[expression] = res              
            return res
        
        memo = {}
        return dfs(expression,memo)